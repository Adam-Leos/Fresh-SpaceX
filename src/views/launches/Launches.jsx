// @flow

import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import LaunchBlock from './LaunchBlock.jsx';
import Spinner from './../../components/spinner/Spinner.jsx';
import './launches.scss';
import type {
  RequestLaunchesDataType,
  ToggleLaunchesModeType,
  ChangeLaunchYearType,
  LaunchesType,
} from '../../flowTypes/flowTypes';

type Props = {
  launches: LaunchesType,
  requestLaunchesData: RequestLaunchesDataType,
  toggleLaunchesMode: ToggleLaunchesModeType,
  changeLaunchYear: ChangeLaunchYearType,
}

type State = {
  isLatestLaunch: boolean,
}

class Launches extends Component<Props, State> {
  constructor() {
    super();

    this.state = {
      isLatestLaunch: true,
    };
  }

  componentDidMount() {
    this.onLaunchesMount();
  }

  onLaunchesMount = async () => {
    if (this.props.launches.launchesData.length === 0) {
      this.props.requestLaunchesData();
    }
  }

  onLaunchesRadioChange = (event: SyntheticInputEvent<HTMLInputElement>) => {
    this.props.toggleLaunchesMode();
  }

  onLaunchYearChange = (event: SyntheticInputEvent<HTMLInputElement>) => {
    const newEnteredYear: string = event.target.value.replace(/[^0-9.]/g, '');
    this.props.changeLaunchYear(newEnteredYear);
  }

  onSubmit = (event: SyntheticInputEvent<HTMLInputElement>) => {
    event.preventDefault();

    let params = this.props.launches.isPastLaunches ? '' : '/upcoming';

    params += `?launch_year=${this.props.launches.launchYear}`;
    this.props.requestLaunchesData(params);

    this.setState({
      isLatestLaunch: false,
    });
  }

  render() {
    const {
      isPastLaunches,
      launchesData,
      launchYear,
      error,
    } = this.props.launches;
    const { isLatestLaunch } = this.state;

    return (
      <div className="launches">
        <form action="" onSubmit={this.onSubmit} className="launches__form">
          <label htmlFor="past-launches-radio" className="launches__label">
            I want to see the past (past launches)
            <input
              type="radio"
              id="past-launches-radio"
              name="launch-mode"
              value="latest"
              checked={isPastLaunches}
              onChange={this.onLaunchesRadioChange}
              className="launches__radio"
            />
          </label>
          <label htmlFor="upcoming-launches-radio" className="launches__label">
            I want to see the future (upcoming launches)
            <input
              type="radio"
              id="upcoming-launches-radio"
              name="launch-mode"
              value="upcoming"
              checked={!isPastLaunches}
              onChange={this.onLaunchesRadioChange}
              className="launches__radio"
            />
          </label>
          <label htmlFor="launch-year" className="launches__label">
            Enter Year
            <input
              type="text"
              id="launch-year"
              name="launch-year"
              value={launchYear}
              onChange={this.onLaunchYearChange}
              className="launches__input"
            />
          </label>
          <button type="submit" className="launches__submit">Submit</button>
        </form>
        <h2 className="launches__results-heading">{isLatestLaunch ? 'Latest Launch' : `Year: ${launchYear}` }</h2>
        <div className="launches__grid">
          {(() => {
            if (error !== undefined) {
              return <div>{error.toString()}</div>;
            } else if (launchesData.length === 0) {
              return <Spinner />;
            }

            return launchesData.map((launchData, index) => {
              const flightNumber = launchData.flight_number;
              const linkLocation = {
                pathname: `/launches/${flightNumber}`,
                state: { launchData },
              };

              return (
                <Link to={linkLocation} key={flightNumber} className="launches__link">
                  <LaunchBlock {...launchData} />
                </Link>
              );
            });
          })()}
        </div>
      </div>
    );
  }
}

export default Launches;
