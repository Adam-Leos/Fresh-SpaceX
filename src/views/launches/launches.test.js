import React from 'react';
import Launches from './Launches.jsx';

describe('Testing launches component', () => {
  const toggleLaunchesModeMock = jest.fn()
  const defaultProps = {
    launches: {
      launchesData: [],
      isPastLaunches: true,
      launchYear: '2018',
    },
    toggleLaunchesMode: toggleLaunchesModeMock,
  };

  test('Launches is matching snapshot', () => {
    const component = shallow(<Launches {...defaultProps} />);

    expect(component).toMatchSnapshot();
  });

  test('Laucnhes mode toggling', () => {
    const component = shallow(<Launches {...defaultProps} />);

    expect(component.find('.launches__radio')).toHaveLength(2);
    expect(component.find('#past-launches-radio').props().checked).toEqual(true);
    expect(component.find('#upcoming-launches-radio').props().checked).toEqual(false);

    component.find('#upcoming-launches-radio').simulate('change');
    expect(toggleLaunchesModeMock.mock.calls.length).toEqual(1);

    expect(component).toMatchSnapshot();
  });
});
