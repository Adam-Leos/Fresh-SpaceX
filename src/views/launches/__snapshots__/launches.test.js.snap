// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing launches component Laucnhes mode toggling 1`] = `
<div
  className="launches"
>
  <form
    action=""
    className="launches__form"
    onSubmit={[Function]}
  >
    <label
      className="launches__label"
      htmlFor="past-launches-radio"
    >
      I want to see the past (past launches)
      <input
        checked={true}
        className="launches__radio"
        id="past-launches-radio"
        name="launch-mode"
        onChange={[Function]}
        type="radio"
        value="latest"
      />
    </label>
    <label
      className="launches__label"
      htmlFor="upcoming-launches-radio"
    >
      I want to see the future (upcoming launches)
      <input
        checked={false}
        className="launches__radio"
        id="upcoming-launches-radio"
        name="launch-mode"
        onChange={[Function]}
        type="radio"
        value="upcoming"
      />
    </label>
    <label
      className="launches__label"
      htmlFor="launch-year"
    >
      Enter Year
      <input
        className="launches__input"
        id="launch-year"
        name="launch-year"
        onChange={[Function]}
        type="text"
        value="2018"
      />
    </label>
    <button
      className="launches__submit"
      type="submit"
    >
      Submit
    </button>
  </form>
  <h2
    className="launches__results-heading"
  >
    Latest Launch
  </h2>
  <div
    className="launches__grid"
  >
    <Spinner />
  </div>
</div>
`;

exports[`Testing launches component Launches is matching snapshot 1`] = `
<div
  className="launches"
>
  <form
    action=""
    className="launches__form"
    onSubmit={[Function]}
  >
    <label
      className="launches__label"
      htmlFor="past-launches-radio"
    >
      I want to see the past (past launches)
      <input
        checked={true}
        className="launches__radio"
        id="past-launches-radio"
        name="launch-mode"
        onChange={[Function]}
        type="radio"
        value="latest"
      />
    </label>
    <label
      className="launches__label"
      htmlFor="upcoming-launches-radio"
    >
      I want to see the future (upcoming launches)
      <input
        checked={false}
        className="launches__radio"
        id="upcoming-launches-radio"
        name="launch-mode"
        onChange={[Function]}
        type="radio"
        value="upcoming"
      />
    </label>
    <label
      className="launches__label"
      htmlFor="launch-year"
    >
      Enter Year
      <input
        className="launches__input"
        id="launch-year"
        name="launch-year"
        onChange={[Function]}
        type="text"
        value="2018"
      />
    </label>
    <button
      className="launches__submit"
      type="submit"
    >
      Submit
    </button>
  </form>
  <h2
    className="launches__results-heading"
  >
    Latest Launch
  </h2>
  <div
    className="launches__grid"
  >
    <Spinner />
  </div>
</div>
`;
